bee-message/src/output/alias.rs

```diff
@@ -296,6 +298,141 @@ impl AliasOutput {
             .map(|unlock_condition| unlock_condition.address())
             .unwrap()
     }
+
+    ///
+    #[inline(always)]
+    pub fn chain_id(&self) -> ChainId {
+        ChainId::Alias(self.alias_id)
+    }
+
+    ///
+    pub fn unlock(
+        &self,
+        output_id: &OutputId,
+        unlock_block: &UnlockBlock,
+        inputs: &[(OutputId, &Output)],
+        context: &mut ValidationContext,
+    ) -> Result<(), ConflictReason> {
+        let alias_id = if self.alias_id().is_null() {
+            AliasId::new(output_id.hash())
+        } else {
+            *self.alias_id()
+        };
+        let next_state = context.output_chains.get(&ChainId::from(alias_id));
+
+        let locked_address = match next_state {
+            Some(Output::Alias(next_state)) => {
+                if self.state_index() == next_state.state_index() {
+                    self.governor_address()
+                } else {
+                    self.state_controller_address()
+                }
+            }
+            Some(_) => unreachable!(),
+            None => self.governor_address(),
+        };
+
+        locked_address.unlock(unlock_block, inputs, context)?;
+
+        context
+            .unlocked_addresses
+            .insert(Address::from(AliasAddress::from(alias_id)));
+
+        Ok(())
+    }
+}
+
+impl StateTransition for AliasOutput {
+    fn creation(next_state: &Self, context: &ValidationContext) -> Result<(), StateTransitionError> {
+        if !next_state.alias_id.is_null() {
+            return Err(StateTransitionError::NonZeroCreatedId);
+        }
+
+        if next_state.state_index != 0 {
+            return Err(StateTransitionError::NonZeroCreatedStateIndex);
+        }
+
+        if next_state.foundry_counter != 0 {
+            return Err(StateTransitionError::NonZeroCreatedFoundryCounter);
+        }
+
+        if let Some(issuer) = next_state.immutable_feature_blocks().issuer() {
+            if !context.unlocked_addresses.contains(issuer.address()) {
+                return Err(StateTransitionError::IssuerNotUnlocked);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn transition(
+        current_state: &Self,
+        next_state: &Self,
+        context: &ValidationContext,
+    ) -> Result<(), StateTransitionError> {
```

Comment:
I think it would be cool to split up these big transition functions into smaller sub-functions with descriptive names. But that doesn't have to be done in this PR.

Reply:
I'd prefer to keep it as one logical function as this is a trait but what I can do is put more comments in it to separate and explain the checks.

Reply:
I'll do that when the TIPs are fully stable and I'll copy paste the checks at the relevant places.